Split sharedialog code

rullzer opened the pull request:
In preparation of the full desktop sharing (https://github.com/owncloud/client/issues/3737) the sharedialog code had to be somewhat unentangled.

There now is more code in libsync and the gui is cleaned up a bit.

* OCSJob class added. To allow a nice abstraction also for the new sharee endpoint code (which is required)
* OcsShareJob class is now in libsync and has more generic functions like createShare, setPassword etc. So the GUI part can just say do this.
* ShareDialog is cleaned up

My tests show that everything still works as expected but please run your own tests to make sure I did not break anything!

CC: @dragotin @guruz @ogoffart and others that want to test!

phil-davis commented on src/libsync/ocssharejob.h:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Set the password of a share
+     *
+     * @param password The apssword of the share, if the password is empty the
```
s/apssword/password

ogoffart commented:
Ideally, my long term goal would be that libsync would be for the sync engine.
Adding these jobs there goes against my goal.

rullzer commented:
@ogoffart ah mm, then we could also move the jobs to the gui for now... 
but also they are not really gui... yet is not ideal...

I'll move them to the gui for now.

guruz commented:
Not sure I agree with @ogoffart .. sharing will be more prominent in future.. hm

dragotin commented:
Not agreeing with @ogoffart is hardly an option.

rullzer commented:
So finally my qt5 setup is running :)
All the files are now back in gui. But we might want to move to an libocs or libshare in the not to distant future. But I need somebody with some cmake-fu for that.

Please tests and review :)

phil-davis commented on src/gui/ocssharejob.h:
```diff
@@ -0,0 +1,90 @@
+    explicit OcsShareJob(AccountPtr account, QObject *parent = 0);
+
+    /**
+     * Constructors for exisiting shares of which we know the shareId
```
s/exisitng/existing/

phil-davis commented on src/gui/ocssharejob.h:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Get all the shares
+     *
+     * @param path Path to requests shares for (default all shares)
```
s/requests/request/

phil-davis commented on src/gui/ocsjob.h:
```diff
@@ -0,0 +1,115 @@
+
+}
+
+#endif // OCSSHAREJOB_H
```
OCSJOB_H

rullzer commented:
@phil-davis thanks. Fixed

guruz commented:
I'm not working on oC these days, maybe @danimo @dragotin can have a look.

dragotin commented on src/gui/ocssharejob.cpp:
```diff
@@ -0,0 +1,106 @@
+    QList<QPair<QString, QString> > getParams;
+    getParams.append(qMakePair(QString::fromLatin1("path"), path));
+    setGetParams(getParams);
+
```
I think an API that provides a method `addParam( QPair())` instead of the `setGetParams(QList())` would be nicer because it would save to set up the List and all the pairing here. Could we also use just an `addParam()`  rather than addGetParam() and `addPostParam()`?

rullzer replied:
Yes I like that. let me take care of that.

dragotin commented:
Apart from my comment above (which is kind of minor but maybe still nice to have) I like this very much  :+1:

As usual my question: Do we see any chance to (unit-) test our job implementations, especially with a focus on error handling that we do? Any idea?

ckamm commented:
@rullzer Feel free to defer my comment about splitting this into several classes and result parsing to later. I think this is already way nicer right now! :)

rullzer commented:
Fixes are on their way. Basically what I noticed is that yet another layer of abstraction would be nice. So that we have share objects. 

Then the gui interacts with a share object (sets password etc). The object in the background then makes sure that the proper jobs are fired etc. That should make the gui code even cleaner and allow us to have really proper separation between the logic and the gui. But also that is future work.

rullzer commented:
Ok pushed. Lots of commits (but squashing will come later).
Code is a lot cleaner now. But we still have room for improvement.

@ckamm I addressed most of your comments I think. Thanks.

@ckamm @dragotin Could you have another look before I start squashing commits and merge it.

phil-davis commented on src/gui/ocsjob.h:
```diff
@@ -0,0 +1,122 @@
+ * @ingroup gui
+ *
+ * Base class for jobs that talk to the OCS endpoints on the server.
+ * All the comminication logic is handled in this class.
```
s/comminication/communication/
I will just add low-value comments by finding the comment typos, and along the way maybe some day be able to learn how this beast all fits together and then contribute something actually useful :)

phil-davis commented on src/gui/ocsjob.h:
```diff
@@ -0,0 +1,122 @@
+    void addPassStatusCode(int code);
+
+    /**
+     * The base path for an OcsJob is always the same. But it could be that case that
```
"that case that" => "the case that"

phil-davis commented on src/gui/ocsjob.h:
```diff
@@ -0,0 +1,122 @@
+     * The base path for an OcsJob is always the same. But it could be that case that
+     * certain operations need to append something to the URL.
+     *
+     * This functions appends the common id. so <PATH>/<ID>
```
s/functions/function/

phil-davis commented on src/gui/ocssharejob.h:
```diff
@@ -0,0 +1,107 @@
+    void setPassword(int shareId, const QString& password);
+
+    /**
+     * Void set the a share to be public upload
```
"the a" - pick just one of these.

phil-davis commented on src/gui/thumbnailjob.h:
```diff
@@ -0,0 +1,51 @@
+ * @ingroup gui
+ *
+ * Job that allows fetching a preview (of 150x150 for now) of a given file.
+ * Once the job finished the jobFinished signal will be emitted.
```
"job has finished"

ckamm commented:
@rullzer The fixes so far look good. Unfortunately I won't have time for another full review next week.

rullzer commented:
Thanks @ckamm @phil-davis

guruz commented on src/gui/ocssharejob.h:
```diff
@@ -0,0 +1,107 @@
+        CREATE = 4,
+        DELETE = 8,
+        SHARE = 16,
+        ALL = 31
```
Why not READ | UPDATE | ... ?

ogoffart commented on src/gui/ocssharejob.h:
```diff
@@ -0,0 +1,107 @@
+    /**
+     * Support sharetypes
+     */
+    enum class SHARETYPE : int {
```
is enum class  supported on the minimum compiler we support?
